### ISSUE TITLE
Funkcije gubitka - debuggiranje

### ISSUE THREAD

**Issue by LeoPlese** [eclipse907/Projekt#5]
Bug kod svih funkcija gubitka su jednog od dva tipa:
-CrossEntropyLoss - gubitak već nakon nekoliko iteracija postane "inf" jer je log(mali broj blizu 0) = inf
-HingeLoss - slično kao i CrossEntropyLoss, gubitak naglo raste dok ne dođe do "inf" nakon otprilike 90 iteracija kad postaoje NaN
-L1Loss, L1SmoothLoss, L2Loss - gubitak je najbolji u 0. iteraciji, zatim nekoliko iteracija pomalo raste i onda ostane na toj vrijednosti do kraja

U slučaju svake primijenjene funkcije gubitka gubitak znači raste umjesto da pada ili se pak stabilizira na jednoj vrijednosti.

Kod za debuggirati nalazi se u grani l2regularizer.

Napomena: regularizatore također trebamo debuggirati, a to bismo nakon što debuggiramo funkcije gubitka i utvrdimo da one ispravno rade.

**Comment by ssegvic** [comment]
Zvuči kao da se gradijenti gubitka krivo računaju. Navedite upute kako reproducirati pogrešku.

**Comment by LeoPlese** [comment]
> Zvuči kao da se gradijenti gubitka krivo računaju. Navedite upute kako reproducirati pogrešku.

Testirali smo na:
- N=10 i C=10 - postavljaju se na početku main funkcije u fcann2.py
- niter = 100 - postavlja se u paramaters.py
- za ime modula u kojem se nalazi funkcija gubitka (testirano na svima) - postavlja se na početku main funkcije u fcann2.py:
lossModule = import_module("losses.CrossEntropyLoss")
ili
lossModule = import_module("losses.L2Loss")
ili
lossModule = import_module("losses.L1Loss")
ili
lossModule = import_module("losses.L1SmoothLoss")
ili
lossModule = import_module("losses.HingeLoss")
- odgovor na "Da li želite koristiti rano zaustavljanje: " staviti na "ne" (za "da" uvijek daje optimalni niter = 1 jer se pogreška za svaki korišteni loss poveća već u sljedećij iteraciji tj. unutar broja iteracija određenih parametrom "patience" u paramaters.py se ne pronađe manja greška)
- upis regularizatora za korištenje je privremeno zakomentiran jer se želimo fokusirati prvo debuggirati funkcije gubitka
- upis modula optimizatora i gradchecka je zakomentiran jer se fokusiramo na debuggiranje funkcija gubitka

Sve ove stvari su namještene u grani l2regularizer i prisutne u verziji na commitu "remove commit" hasha 535ed0f1d8e78eff1bd223b471ec8e1524b0fe27
tako da se tamo treba samo mijenjati ime modula funkcije gubitka i eventualno ostali parametri.

Hvala Vam unaprijed na pomoći!

**Comment by ssegvic** [comment]
Došlo je do nesporazuma između optimizatora i glavnog programa.

Optimizator kaže:
```
    def __sgd(self, grad_W1, grad_W2):
        grad_W1 = np.transpose(self.model.W1) - self.params.learning_rate_sgd * grad_W1
        grad_W2 = np.transpose(self.model.W2) - self.params.learning_rate_sgd * grad_W2
        return grad_W1, grad_W2
```

Glavni program kaže (funkcija `train`):
```
       grad_W1, grad_W2 = optimizationClass(grad_W1, grad_W2)
       # ...
       model.W1 += np.transpose(grad_W1)
```
Optimizator je već izračunao novu vrijednost parametra, a glavni program bi je trebao samo preuzeti. Funkciju `train` bi trebalo popraviti na sljedeći način:
```
       new_params = optimizationClass(model.get_params())
       model.set_params(new_params)
```
Optimizator bi trebalo popraviti tako da ne prima `W1` i `W2`, nego listu svih parametara. Želimo imati takav optimizator zato što bi se on mogao primijeniti na bilo kakav model.

**Comment by LeoPlese** [comment]
Oprostite Profesore, jeste li možda naišli što bi moglo biti krivo s računanjem gradijenta u našim Loss klasama?

Što se tiče optimizacije, dok smo testirali funkcije gubitka, optimizaciju nismo pozivali tako da je bug vjerojatno u računanju gradijenta kako ste rekli.

Hvala Vam unaprijed!

**Comment by ssegvic** [comment]
Mislim da su gradijenti točni.

**Comment by ssegvic** [comment]
Gubitak dobro izračuna dL_ds2, a model taj tenzor ispravno prevodi u gradijente po parametrima.
